pubsublite/internal/wire/publisher.go

```diff
@@ -0,0 +1,260 @@
+	batcher            *publishMessageBatcher
+	enableSendToStream bool
+
+	abstractService
+}
+
+// singlePartitionPublisherFactory creates instances of singlePartitionPublisher
+// for given partition numbers.
+type singlePartitionPublisherFactory struct {
+	ctx       context.Context
+	pubClient *vkit.PublisherClient
+	settings  PublishSettings
+	topicPath string
+}
+
+func (f *singlePartitionPublisherFactory) New(partition int) *singlePartitionPublisher {
+	pp := &singlePartitionPublisher{
+		pubClient: f.pubClient,
+		topic:     topicPartition{Path: f.topicPath, Partition: partition},
+		initialReq: &pb.PublishRequest{
+			RequestType: &pb.PublishRequest_InitialRequest{
+				InitialRequest: &pb.InitialPublishRequest{
+					Topic:     f.topicPath,
+					Partition: int64(partition),
+				},
+			},
+		},
+	}
+	pp.batcher = newPublishMessageBatcher(&f.settings, partition, pp.onNewBatch)
+	pp.stream = newRetryableStream(f.ctx, pp, f.settings.Timeout, reflect.TypeOf(pb.PublishResponse{}))
+	return pp
+}
+
+// Start attempts to establish a publish stream connection.
+func (pp *singlePartitionPublisher) Start() {
+	pp.mu.Lock()
+	defer pp.mu.Unlock()
+
+	if pp.unsafeUpdateStatus(serviceStarting, nil) {
+		pp.stream.Start()
+	}
+}
+
+// Stop initiates shutdown of the publisher. All pending messages are flushed.
+func (pp *singlePartitionPublisher) Stop() {
+	pp.mu.Lock()
+	defer pp.mu.Unlock()
+	pp.unsafeInitiateShutdown(serviceTerminating, nil)
+}
+
+// Publish a pub/sub message.
+func (pp *singlePartitionPublisher) Publish(msg *pb.PubSubMessage, onResult PublishResultFunc) {
+	pp.mu.Lock()
+	defer pp.mu.Unlock()
+
+	processMessage := func() error {
+		if err := pp.unsafeCheckServiceStatus(); err != nil {
+			return err
+		}
+		if err := pp.batcher.AddMessage(msg, onResult); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// If the new message cannot be published, flush pending messages and then
+	// terminate the stream once results are received.
+	if err := processMessage(); err != nil {
+		pp.unsafeInitiateShutdown(serviceTerminating, err)
+		onResult(nil, err)
+	}
+}
+
+func (pp *singlePartitionPublisher) newStream(ctx context.Context) (grpc.ClientStream, error) {
+	return pp.pubClient.Publish(addTopicRoutingMetadata(ctx, pp.topic))
+}
+
+func (pp *singlePartitionPublisher) initialRequest() (interface{}, bool) {
+	return pp.initialReq, true
+}
+
+func (pp *singlePartitionPublisher) validateInitialResponse(response interface{}) error {
+	pubResponse, _ := response.(*pb.PublishResponse)
+	if pubResponse.GetInitialResponse() == nil {
+		return errInvalidInitialPubResponse
+	}
+	return nil
+}
+
+func (pp *singlePartitionPublisher) onStreamStatusChange(status streamStatus) {
+	pp.mu.Lock()
+	defer pp.mu.Unlock()
+
+	switch status {
+	case streamReconnecting:
+		// Prevent onNewBatch() from sending any new batches to the stream.
+		pp.enableSendToStream = false
+
+	case streamConnected:
+		pp.unsafeUpdateStatus(serviceActive, nil)
```

Comment:
won't this leave it as active if sending a message to the stream fails? consider instead:

```golang
batches := pp.batcher.InFlightBatches()
		for _, batch := range batches {
			if !pp.stream.Send(batch.ToPublishRequest()) return;
		}

pp.unsafeUpdateStatus(serviceActive, nil)
pp.enableSendToStream = true
```

Reply:
Done.

Except I left `pp.unsafeUpdateStatus(serviceActive, nil)` at the beginning. It just transitions the publisher to the `serviceActive` state (duplicates ignored), which means the underlying stream is connected or reconnecting, so it's ok to leave it as active. The publisher transitions to `serviceTerminating|serviceTerminated` on shutdown.